src/runtime/dict.cpp

```diff
@@ -596,21 +596,33 @@ extern "C" Box* dictNew(Box* _cls, BoxedTuple* args, BoxedDict* kwargs) {
     return new (cls) BoxedDict();
 }
 
-void dictMerge(BoxedDict* self, Box* other) {
+bool dictMergeImpl(BoxedDict* self, Box* other) {
     if (PyDict_Check(other)) {
         for (const auto& p : static_cast<BoxedDict*>(other)->d)
             self->d[p.first] = p.second;
-        return;
+        return true;
     }
 
     static BoxedString* keys_str = internStringImmortal("keys");
     CallattrFlags callattr_flags{.cls_only = false, .null_on_nonexistent = true, .argspec = ArgPassSpec(0) };
     Box* keys = callattr(other, keys_str, callattr_flags, NULL, NULL, NULL, NULL, NULL);
-    assert(keys);
+    if (!keys) {
```

Comment:
So the reason this was an assert is because CPython actually does something different in this case -- it calls PyDict_MergeFromSeq2 -- and we shouldn't just ignore that behavior.  Can we copy in that function and then call it if there was no "keys" attribute?